src/opt/ReplaceHalfMaskedMemOps.cpp

```diff
@@ -0,0 +1,216 @@
+    if (auto *CV = llvm::dyn_cast<llvm::ConstantVector>(mask)) {
+        auto N = CV->getType()->getNumElements();
+        for (auto i = N / 2; i < N; i++) {
+            llvm::Constant *E = CV->getAggregateElement(i);
+            if (!E || !llvm::isa<llvm::ConstantInt>(E) || !llvm::cast<llvm::ConstantInt>(E)->isZero()) {
+                // found non-zero element
+                return false;
+            }
+        }
+        for (auto i = 0; i < N / 2; i++) {
+            llvm::Constant *E = CV->getAggregateElement(i);
+            if (!E || !llvm::isa<llvm::ConstantInt>(E) || !llvm::cast<llvm::ConstantInt>(E)->isOne()) {
+                // found non-one element
+                return false;
+            }
+        }
+        return true;
+    }
+    return false;
+}
+
+llvm::Value *lMergeVectors(llvm::IRBuilder<> &B, llvm::Value *firstVector, llvm::Value *secondVector,
+                           llvm::Twine &name) {
+    auto *firstVecType = llvm::cast<llvm::VectorType>(firstVector->getType());
+    auto *secondVecType = llvm::cast<llvm::VectorType>(secondVector->getType());
+    unsigned N = firstVecType->getElementCount().getKnownMinValue();
+
+    assert(firstVecType == secondVecType);
+
+    // Create a vector of indices to merge two vectors
+    std::vector<llvm::Constant *> indices;
+    for (unsigned i = 0; i < N * 2; ++i) {
+        indices.push_back(llvm::ConstantInt::get(B.getInt32Ty(), i));
+    }
+    llvm::ArrayRef<llvm::Constant *> indicesRef(indices);
+    llvm::Constant *mask = llvm::ConstantVector::get(indicesRef);
+
+    // Use shufflevector instruction to create the new vector
+    return B.CreateShuffleVector(firstVector, secondVector, mask, name);
+}
+
+llvm::Value *lShrinkVector(llvm::IRBuilder<> &B, llvm::Value *originalVector) {
+    auto *vecType = llvm::cast<llvm::VectorType>(originalVector->getType());
+    unsigned N = vecType->getElementCount().getKnownMinValue();
+
+    // Create a vector of indices to select the first half of the elements
+    std::vector<llvm::Constant *> indices;
+    for (unsigned i = 0; i < N / 2; ++i) {
+        indices.push_back(llvm::ConstantInt::get(B.getInt32Ty(), i));
+    }
+    llvm::ArrayRef<llvm::Constant *> indicesRef(indices);
+    llvm::Constant *mask = llvm::ConstantVector::get(indicesRef);
+
+    // Use shufflevector instruction to create the new vector
+    llvm::Twine name = originalVector->getName() + ".half";
+    return B.CreateShuffleVector(originalVector, llvm::UndefValue::get(vecType), mask, name);
+}
+
+llvm::Value *lBitcastPointerType(llvm::IRBuilder<> &B, llvm::Value *ptr, llvm::Value *value) {
+    auto *vecType = llvm::cast<llvm::VectorType>(value->getType());
+    auto *newPtrType = llvm::PointerType::get(vecType, 0 /* TODO! */);
+    // TODO! opaque pointer is no-op here, any special handling?
+    return B.CreateBitCast(ptr, newPtrType);
+}
+
+llvm::Constant *lShrinkConstVec(llvm::LLVMContext &context, llvm::Value *originalValue) {
```

Comment:
It would clearer if the method accepted `llvm::Constant*` argument.

Reply:
I agree, fixed